Deal with OS-dependent line endings in tests

See #101 for details. Short version - some tests fail on Windows (on CI, weirdly not a problem locally) due to line endings. We're now chomping lines to get around this, but it would be good to deal with it a bit more gracefully.

Test output on Windows from this:

```julia
function equivalent(x::AbstractString, y::AbstractString)
    @show x
    @show y
    while endswith(x, "\n")
        x = x[1:end-1] # trailing newline characters are ambiguous
    end
    while endswith(y, "\n")
        y = y[1:end-1]
    end
    x == y
end
 
⋮
yaml_string = read(yaml_file_name, String)
⋮
```

```
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
x = "Mark McGwire"
y = "Mark McGwire"
x = "65 Home Runs\n0.278 Batting Average\n"
y = "65 Home Runs\n0.278 Batting Average\n"
x = "Mark set a major league home run record in 1998.\n"
y = "Mark set a major league home run record in 1998.\n"
Test Summary:     | Pass  Total  Time
test = spec-02-16 |    9      9  0.0s
```

and fails

```
Test Summary:     | Pass  Total  Time
test = spec-02-23 |    9      9  0.3s
Load All from String: Error During Test at C:\Users\tanak\git\YAML.jl\test\runtests.jl:204
  Test threw exception
  Expression: begin
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:205 =#
    data = YAML.load_all(yaml_string, TestConstructor())
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:209 =#
    equivalent(first(data), expected)
end
  AssertionError: forward!(composer.input) isa DocumentStartEvent
  Stacktrace:
    [1] compose_document(composer::YAML.Composer)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:44
    [2] compose(events::YAML.EventStream, resolver::YAML.Resolver)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:33
    [3] load(tokenstream::YAML.TokenStream, constructor::YAML.Constructor)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:69
    [4] next(it::YAML.YAMLDocIterator, state::Nothing)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:112
    [5] iterate
      @ C:\Users\tanak\git\YAML.jl\src\YAML.jl:120 [inlined]
    [6] first
      @ .\abstractarray.jl:472 [inlined]
    [7] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:209 [inlined]
    [8] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:669 [inlined]
    [9] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:204 [inlined]
   [10] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1577 [inlined]
   [11] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:204 [inlined]
   [12] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1669 [inlined]
   [13] top-level scope
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:143
Load All from String: Error During Test at C:\Users\tanak\git\YAML.jl\test\runtests.jl:212
  Test threw exception
  Expression: begin
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:213 =#
    dictData = YAML.load_all(yaml_string, more_constructors, multi_constructors)
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:217 =#
    equivalent(first(dictData), expected)
end
  AssertionError: forward!(composer.input) isa DocumentStartEvent
  Stacktrace:
    [1] compose_document(composer::YAML.Composer)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:44
    [2] compose(events::YAML.EventStream, resolver::YAML.Resolver)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:33
    [3] load(tokenstream::YAML.TokenStream, constructor::YAML.Constructor)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:69
    [4] next(it::YAML.YAMLDocIterator, state::Nothing)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:112
    [5] iterate
      @ C:\Users\tanak\git\YAML.jl\src\YAML.jl:120 [inlined]
    [6] first
      @ .\abstractarray.jl:472 [inlined]
    [7] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:217 [inlined]
    [8] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:669 [inlined]
    [9] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:212 [inlined]
   [10] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1577 [inlined]
   [11] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:204 [inlined]
   [12] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1669 [inlined]
   [13] top-level scope
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:143
Test Summary:          | Pass  Error  Total  Time
test = empty_scalar    |    7      2      9  1.1s
  Load from File       |    2             2  0.0s
  Load from String     |    2             2  0.0s
  Load All from File   |    2             2  0.0s
  Load All from String |           2      2  1.0s
  Writing              |    1             1  0.0s
ERROR: LoadError: Some tests did not pass: 7 passed, 0 failed, 2 errored, 0 broken.
in expression starting at C:\Users\tanak\git\YAML.jl\test\runtests.jl:3
ERROR: Package YAML errored during testing
```

Test output from this:

```
function equivalent(x::AbstractString, y::AbstractString)
    @show x
    @show y
    # while endswith(x, "\n")
    #     x = x[1:end-1] # trailing newline characters are ambiguous
    # end
    # while endswith(y, "\n")
    #     y = y[1:end-1]
    # end
    x == y
end

⋮
yaml_string = read(yaml_file_name, String)
⋮
```

on WIndows:

```
Test Summary:     | Pass  Total  Time
test = spec-02-23 |    9      9  0.3s
Load All from String: Error During Test at C:\Users\tanak\git\YAML.jl\test\runtests.jl:205
  Test threw exception
  Expression: begin
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:206 =#
    data = YAML.load_all(yaml_string, TestConstructor())
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:210 =#
    equivalent(first(data), expected)
end
  AssertionError: forward!(composer.input) isa DocumentStartEvent
  Stacktrace:
    [1] compose_document(composer::YAML.Composer)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:44
    [2] compose(events::YAML.EventStream, resolver::YAML.Resolver)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:33
    [3] load(tokenstream::YAML.TokenStream, constructor::YAML.Constructor)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:69
    [4] next(it::YAML.YAMLDocIterator, state::Nothing)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:112
    [5] iterate
      @ C:\Users\tanak\git\YAML.jl\src\YAML.jl:120 [inlined]
    [6] first
      @ .\abstractarray.jl:472 [inlined]
    [7] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:210 [inlined]
    [8] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:669 [inlined]
    [9] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:205 [inlined]
   [10] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1577 [inlined]
   [11] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:205 [inlined]
   [12] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1669 [inlined]
   [13] top-level scope
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:144
Load All from String: Error During Test at C:\Users\tanak\git\YAML.jl\test\runtests.jl:213
  Test threw exception
  Expression: begin
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:214 =#
    dictData = YAML.load_all(yaml_string, more_constructors, multi_constructors)
    #= C:\Users\tanak\git\YAML.jl\test\runtests.jl:218 =#
    equivalent(first(dictData), expected)
end
  AssertionError: forward!(composer.input) isa DocumentStartEvent
  Stacktrace:
    [1] compose_document(composer::YAML.Composer)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:44
    [2] compose(events::YAML.EventStream, resolver::YAML.Resolver)
      @ YAML C:\Users\tanak\git\YAML.jl\src\composer.jl:33
    [3] load(tokenstream::YAML.TokenStream, constructor::YAML.Constructor)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:69
    [4] next(it::YAML.YAMLDocIterator, state::Nothing)
      @ YAML C:\Users\tanak\git\YAML.jl\src\YAML.jl:112
    [5] iterate
      @ C:\Users\tanak\git\YAML.jl\src\YAML.jl:120 [inlined]
    [6] first
      @ .\abstractarray.jl:472 [inlined]
    [7] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:218 [inlined]
    [8] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:669 [inlined]
    [9] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:213 [inlined]
   [10] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1577 [inlined]
   [11] macro expansion
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:205 [inlined]
   [12] macro expansion
      @ C:\Users\tanak\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\share\julia\stdlib\v1.10\Test\src\Test.jl:1669 [inlined]
   [13] top-level scope
      @ C:\Users\tanak\git\YAML.jl\test\runtests.jl:144
Test Summary:          | Pass  Error  Total  Time
test = empty_scalar    |    7      2      9  1.1s
  Load from File       |    2             2  0.0s
  Load from String     |    2             2  0.0s
  Load All from File   |    2             2  0.0s
  Load All from String |           2      2  1.1s
  Writing              |    1             1  0.0s
ERROR: LoadError: Some tests did not pass: 7 passed, 0 failed, 2 errored, 0 broken.
in expression starting at C:\Users\tanak\git\YAML.jl\test\runtests.jl:3
ERROR: Package YAML errored during testing
```

on macOS:

```
Test Summary:     | Pass  Total  Time
test = spec-02-23 |    9      9  0.3s
Test Summary:       | Pass  Total  Time
test = empty_scalar |    9      9  0.0s
```

--- a/src/YAML.jl
+++ b/src/YAML.jl
@@ -100,6 +100,11 @@ end
 
 YAMLDocIterator(input::IO, more_constructors::_constructor=nothing, multi_constructors::Dict = Dict(); dicttype::_dicttype=Dict{Any, Any}, constructorType::Function = SafeConstructor) = YAMLDocIterator(input, constructorType(_patch_constructors(more_constructors, dicttype), multi_constructors))
 
+# It's unknown how many documents will be found. By doing this,
+# functions like `collect` do not try to query the length of the
+# iterator.
+Base.IteratorSize(::YAMLDocIterator) = Base.SizeUnknown()
+
 # Old iteration protocol:
 start(it::YAMLDocIterator) = nothing
 

--- a/src/YAML.jl
+++ b/src/YAML.jl
@@ -105,10 +105,9 @@ YAMLDocIterator(input::IO, more_constructors::_constructor=nothing, multi_constr
 # iterator.
 Base.IteratorSize(::YAMLDocIterator) = Base.SizeUnknown()
 
-# Old iteration protocol:
-start(it::YAMLDocIterator) = nothing
-
-function next(it::YAMLDocIterator, state)
+# Iteration protocol.
+function iterate(it::YAMLDocIterator, _ = nothing)
+    it.next_doc === nothing && return nothing
     doc = it.next_doc
     if eof(it.input)
         it.next_doc = nothing
@@ -118,12 +117,6 @@ function next(it::YAMLDocIterator, state)
     end
     return doc, nothing
 end
-
-done(it::YAMLDocIterator, state) = it.next_doc === nothing
-
-# 0.7 iteration protocol:
-iterate(it::YAMLDocIterator) = next(it, start(it))
-iterate(it::YAMLDocIterator, s) = done(it, s) ? nothing : next(it, s)
 
 """
     load_all(x::Union{AbstractString, IO}) -> YAMLDocIterator

--- a/src/composer.jl
+++ b/src/composer.jl
@@ -41,6 +41,7 @@ end
 
 
 function compose_document(composer::Composer)
+    peek(composer.input) isa StreamEndEvent && return nothing
     @assert forward!(composer.input) isa DocumentStartEvent
     node = compose_node(composer)
     @assert forward!(composer.input) isa DocumentEndEvent

--- a/src/constructor.jl
+++ b/src/constructor.jl
@@ -56,6 +56,7 @@ function construct_document(constructor::Constructor, node::Node)
     data
 end
 
+construct_document(::Constructor, ::Nothing) = nothing
 
 function construct_object(constructor::Constructor, node::Node)
     if haskey(constructor.constructed_objects, node)

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -168,7 +168,7 @@ const testdir = dirname(@__FILE__)
     yaml_file_name = joinpath(testdir, "yaml/$test.yaml")
     julia_file_name = joinpath(testdir, "julia/$test.jl")
 
-    yaml_string = chomp(read(yaml_file_name, String))
+    yaml_string = read(yaml_file_name, String)
     expected = evalfile(julia_file_name)
 
     @testset "Load from File" begin
@@ -444,10 +444,39 @@ end
     @test_throws YAML.ScannerError YAML.load(""" '''a'' """)
 end
 
+# issue 129 - Comment only content
+@testset "issue129" begin
+    @test YAML.load("#") === nothing
+    @test isempty(YAML.load_all("#"))
+end
+
+# issue 132 - load_all fails on windows
+@testset "issue132" begin
+    input = """
+            ---
+            creator: LAMMPS
+            timestep: 0
+            ...
+            ---
+            creator: LAMMPS
+            timestep: 1
+            ...
+            """
+    expected = [Dict("creator" => "LAMMPS", "timestep" => 0),
+                Dict("creator" => "LAMMPS", "timestep" => 1)]
+    @test collect(YAML.load_all(input)) == expected
+end
+
 # issue #148 - warn unknown directives
 @testset "issue #148" begin
     @test (@test_logs (:warn, """unknown directive name: "FOO" at line 1, column 4. We ignore this.""") YAML.load("""%FOO  bar baz\n\n--- "foo\"""")) == "foo"
     @test (@test_logs (:warn, """unknown directive name: "FOO" at line 1, column 4. We ignore this.""") (:warn, """unknown directive name: "BAR" at line 2, column 4. We ignore this.""") YAML.load("""%FOO\n%BAR\n--- foo""")) == "foo"
+end
+
+# issue #143 - load empty file
+@testset "issue #143" begin
+    @test YAML.load("") === nothing
+    @test isempty(YAML.load_all(""))
 end
 
 # issue #144

